File: OpenPlanter/agent/credentials.py
from __future__ import annotations

import getpass
import json
import os
import stat
import sys
from dataclasses import dataclass, field
from pathlib import Path


@dataclass(slots=True)
class CredentialBundle:
    openai_api_key: str | None = None
    anthropic_api_key: str | None = None
    openrouter_api_key: str | None = None
    cerebras_api_key: str | None = None
    exa_api_key: str | None = None
    voyage_api_key: str | None = None

    def has_any(self) -> bool:
        return bool(
            (self.openai_api_key and self.openai_api_key.strip())
            or (self.anthropic_api_key and self.anthropic_api_key.strip())
            or (self.openrouter_api_key and self.openrouter_api_key.strip())
            or (self.cerebras_api_key and self.cerebras_api_key.strip())
            or (self.exa_api_key and self.exa_api_key.strip())
            or (self.voyage_api_key and self.voyage_api_key.strip())
        )

    def merge_missing(self, other: "CredentialBundle") -> None:
        if not self.openai_api_key and other.openai_api_key:
            self.openai_api_key = other.openai_api_key
        if not self.anthropic_api_key and other.anthropic_api_key:
            self.anthropic_api_key = other.anthropic_api_key
        if not self.openrouter_api_key and other.openrouter_api_key:
            self.openrouter_api_key = other.openrouter_api_key
        if not self.cerebras_api_key and other.cerebras_api_key:
            self.cerebras_api_key = other.cerebras_api_key
        if not self.exa_api_key and other.exa_api_key:
            self.exa_api_key = other.exa_api_key
        if not self.voyage_api_key and other.voyage_api_key:
            self.voyage_api_key = other.voyage_api_key

    def to_json(self) -> dict[str, str]:
        out: dict[str, str] = {}
        if self.openai_api_key:
            out["openai_api_key"] = self.openai_api_key
        if self.anthropic_api_key:
            out["anthropic_api_key"] = self.anthropic_api_key
        if self.openrouter_api_key:
            out["openrouter_api_key"] = self.openrouter_api_key
        if self.cerebras_api_key:
            out["cerebras_api_key"] = self.cerebras_api_key
        if self.exa_api_key:
            out["exa_api_key"] = self.exa_api_key
        if self.voyage_api_key:
            out["voyage_api_key"] = self.voyage_api_key
        return out

    @classmethod
    def from_json(cls, payload: dict[str, str] | None) -> "CredentialBundle":
        if not isinstance(payload, dict):
            return cls()
        return cls(
            openai_api_key=(payload.get("openai_api_key") or "").strip() or None,
            anthropic_api_key=(payload.get("anthropic_api_key") or "").strip() or None,
            openrouter_api_key=(payload.get("openrouter_api_key") or "").strip() or None,
            cerebras_api_key=(payload.get("cerebras_api_key") or "").strip() or None,
            exa_api_key=(payload.get("exa_api_key") or "").strip() or None,
            voyage_api_key=(payload.get("voyage_api_key") or "").strip() or None,
        )


def _strip_quotes(value: str) -> str:
    value = value.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {"'", '"'}:
        return value[1:-1]
    return value


def parse_env_file(path: Path) -> CredentialBundle:
    if not path.exists() or not path.is_file():
        return CredentialBundle()
    try:
        lines = path.read_text(encoding="utf-8", errors="replace").splitlines()
    except OSError:
        return CredentialBundle()

    env: dict[str, str] = {}
    for raw in lines:
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        if line.startswith("export "):
            line = line[len("export ") :].strip()
        if "=" not in line:
            continue
        key, value = line.split("=", 1)
        key = key.strip()
        value = _strip_quotes(value.strip())
        env[key] = value

    return CredentialBundle(
        openai_api_key=(env.get("OPENAI_API_KEY") or env.get("OPENPLANTER_OPENAI_API_KEY") or "").strip() or None,
        anthropic_api_key=(env.get("ANTHROPIC_API_KEY") or env.get("OPENPLANTER_ANTHROPIC_API_KEY") or "").strip()
        or None,
        openrouter_api_key=(env.get("OPENROUTER_API_KEY") or env.get("OPENPLANTER_OPENROUTER_API_KEY") or "").strip()
        or None,
        cerebras_api_key=(env.get("CEREBRAS_API_KEY") or env.get("OPENPLANTER_CEREBRAS_API_KEY") or "").strip()
        or None,
        exa_api_key=(env.get("EXA_API_KEY") or env.get("OPENPLANTER_EXA_API_KEY") or "").strip() or None,
        voyage_api_key=(env.get("VOYAGE_API_KEY") or env.get("OPENPLANTER_VOYAGE_API_KEY") or "").strip() or None,
    )


def credentials_from_env() -> CredentialBundle:
    return CredentialBundle(
        openai_api_key=(
            os.getenv("OPENPLANTER_OPENAI_API_KEY")
            or os.getenv("OPENAI_API_KEY")
            or ""
        ).strip()
        or None,
        anthropic_api_key=(
            os.getenv("OPENPLANTER_ANTHROPIC_API_KEY") or os.getenv("ANTHROPIC_API_KEY") or ""
        ).strip()
        or None,
        openrouter_api_key=(
            os.getenv("OPENPLANTER_OPENROUTER_API_KEY") or os.getenv("OPENROUTER_API_KEY") or ""
        ).strip()
        or None,
        cerebras_api_key=(
            os.getenv("OPENPLANTER_CEREBRAS_API_KEY") or os.getenv("CEREBRAS_API_KEY") or ""
        ).strip()
        or None,
        exa_api_key=(os.getenv("OPENPLANTER_EXA_API_KEY") or os.getenv("EXA_API_KEY") or "").strip() or None,
        voyage_api_key=(os.getenv("OPENPLANTER_VOYAGE_API_KEY") or os.getenv("VOYAGE_API_KEY") or "").strip() or None,
    )


def discover_env_candidates(workspace: Path) -> list[Path]:
    ws = workspace.expanduser().resolve()
    candidates: list[Path] = [
        ws / ".env",
    ]
    seen: set[str] = set()
    unique: list[Path] = []
    for path in candidates:
        key = str(path.resolve()) if path.exists() else str(path)
        if key in seen:
            continue
        seen.add(key)
        unique.append(path)
    return unique


@dataclass(slots=True)
class CredentialStore:
    workspace: Path
    session_root_dir: str = ".openplanter"
    credentials_path: Path = field(init=False)

    def __post_init__(self) -> None:
        self.workspace = self.workspace.expanduser().resolve()
        root = self.workspace / self.session_root_dir
        self.credentials_path = root / "credentials.json"
        root.mkdir(parents=True, exist_ok=True)

    def load(self) -> CredentialBundle:
        if not self.credentials_path.exists():
            return CredentialBundle()
        try:
            payload = json.loads(self.credentials_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return CredentialBundle()
        return CredentialBundle.from_json(payload)

    def save(self, creds: CredentialBundle) -> None:
        payload = creds.to_json()
        self.credentials_path.parent.mkdir(parents=True, exist_ok=True)
        self.credentials_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        try:
            os.chmod(self.credentials_path, stat.S_IRUSR | stat.S_IWUSR)
        except OSError:
            pass


_USER_CONFIG_DIR = Path.home() / ".openplanter"


@dataclass(slots=True)
class UserCredentialStore:
    """User-level credential store at ~/.openplanter/credentials.json."""
    credentials_path: Path = field(init=False)

    def __post_init__(self) -> None:
        self.credentials_path = _USER_CONFIG_DIR / "credentials.json"

    def load(self) -> CredentialBundle:
        if not self.credentials_path.exists():
            return CredentialBundle()
        try:
            payload = json.loads(self.credentials_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return CredentialBundle()
        return CredentialBundle.from_json(payload)

    def save(self, creds: CredentialBundle) -> None:
        payload = creds.to_json()
        self.credentials_path.parent.mkdir(parents=True, exist_ok=True)
        self.credentials_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        try:
            os.chmod(self.credentials_path, stat.S_IRUSR | stat.S_IWUSR)
        except OSError:
            pass


def prompt_for_credentials(
    existing: CredentialBundle,
    force: bool = False,
) -> tuple[CredentialBundle, bool]:
    """Prompt user for credential input when interactive.

    Returns (updated_credentials, changed).
    """
    current = CredentialBundle(
        openai_api_key=existing.openai_api_key,
        anthropic_api_key=existing.anthropic_api_key,
        openrouter_api_key=existing.openrouter_api_key,
        cerebras_api_key=existing.cerebras_api_key,
        exa_api_key=existing.exa_api_key,
        voyage_api_key=existing.voyage_api_key,
    )

    should_prompt = force or not current.has_any()
    if not should_prompt:
        return current, False
    if not sys.stdin.isatty():
        return current, False

    if force:
        print("Key configuration mode: press Enter to keep existing values.")
    else:
        print("No API keys configured. Enter keys to enable providers (leave blank to skip).")

    changed = False

    def _ask(label: str, existing_value: str | None) -> str | None:
        nonlocal changed
        if force and existing_value:
            prompt = f"{label} API key [press Enter to keep existing]"
        else:
            prompt = f"{label} API key"
        value = getpass.getpass(prompt + ": ").strip()
        if not value:
            return existing_value
        changed = changed or (value != (existing_value or ""))
        return value

    current.openai_api_key = _ask("OpenAI", current.openai_api_key)
    current.anthropic_api_key = _ask("Anthropic", current.anthropic_api_key)
    current.openrouter_api_key = _ask("OpenRouter", current.openrouter_api_key)
    current.cerebras_api_key = _ask("Cerebras", current.cerebras_api_key)
    current.exa_api_key = _ask("Exa", current.exa_api_key)
    current.voyage_api_key = _ask("Voyage", current.voyage_api_key)
    if not force and current.has_any() and not existing.has_any():
        changed = True
    return current, changed

File: OpenPlanter/openplanter-desktop/crates/op-core/src/credentials.rs
use std::collections::HashMap;
use std::env;
use std::fs;
#[cfg(unix)]
use std::os::unix::fs::PermissionsExt;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

/// A bundle of API keys for all supported providers.
///
/// Mirrors the Python `CredentialBundle` dataclass.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CredentialBundle {
    pub openai_api_key: Option<String>,
    pub anthropic_api_key: Option<String>,
    pub openrouter_api_key: Option<String>,
    pub cerebras_api_key: Option<String>,
    pub exa_api_key: Option<String>,
    pub voyage_api_key: Option<String>,
}

impl CredentialBundle {
    /// Returns `true` if any key has a non-empty value.
    pub fn has_any(&self) -> bool {
        let keys: [&Option<String>; 6] = [
            &self.openai_api_key,
            &self.anthropic_api_key,
            &self.openrouter_api_key,
            &self.cerebras_api_key,
            &self.exa_api_key,
            &self.voyage_api_key,
        ];
        keys.iter().any(|k| {
            k.as_ref()
                .map(|v| !v.trim().is_empty())
                .unwrap_or(false)
        })
    }

    /// Fill in missing keys from `other`.
    pub fn merge_missing(&mut self, other: &CredentialBundle) {
        macro_rules! fill {
            ($field:ident) => {
                if self.$field.is_none() {
                    self.$field = other.$field.clone();
                }
            };
        }
        fill!(openai_api_key);
        fill!(anthropic_api_key);
        fill!(openrouter_api_key);
        fill!(cerebras_api_key);
        fill!(exa_api_key);
        fill!(voyage_api_key);
    }

    /// Serialize to JSON map, omitting `None` values.
    pub fn to_json(&self) -> HashMap<String, String> {
        let mut out = HashMap::new();
        macro_rules! add {
            ($field:ident, $key:expr) => {
                if let Some(ref v) = self.$field {
                    out.insert($key.to_string(), v.clone());
                }
            };
        }
        add!(openai_api_key, "openai_api_key");
        add!(anthropic_api_key, "anthropic_api_key");
        add!(openrouter_api_key, "openrouter_api_key");
        add!(cerebras_api_key, "cerebras_api_key");
        add!(exa_api_key, "exa_api_key");
        add!(voyage_api_key, "voyage_api_key");
        out
    }

    /// Deserialize from a JSON map.
    pub fn from_json(payload: &HashMap<String, serde_json::Value>) -> Self {
        fn get_str(map: &HashMap<String, serde_json::Value>, key: &str) -> Option<String> {
            map.get(key)
                .and_then(|v| v.as_str())
                .map(|s| s.trim().to_string())
                .filter(|s| !s.is_empty())
        }
        Self {
            openai_api_key: get_str(payload, "openai_api_key"),
            anthropic_api_key: get_str(payload, "anthropic_api_key"),
            openrouter_api_key: get_str(payload, "openrouter_api_key"),
            cerebras_api_key: get_str(payload, "cerebras_api_key"),
            exa_api_key: get_str(payload, "exa_api_key"),
            voyage_api_key: get_str(payload, "voyage_api_key"),
        }
    }
}

/// Strip surrounding quotes from a value.
fn strip_quotes(s: &str) -> &str {
    let trimmed = s.trim();
    if trimmed.len() >= 2 {
        let first = trimmed.as_bytes()[0];
        let last = trimmed.as_bytes()[trimmed.len() - 1];
        if first == last && (first == b'\'' || first == b'"') {
            return &trimmed[1..trimmed.len() - 1];
        }
    }
    trimmed
}

/// Parse a `.env` file and extract credential keys.
pub fn parse_env_file(path: &Path) -> CredentialBundle {
    let content = match fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return CredentialBundle::default(),
    };

    let mut env_map: HashMap<String, String> = HashMap::new();
    for raw in content.lines() {
        let line = raw.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let line = line.strip_prefix("export ").unwrap_or(line).trim();
        if let Some((key, value)) = line.split_once('=') {
            let key = key.trim();
            let value = strip_quotes(value.trim());
            env_map.insert(key.to_string(), value.to_string());
        }
    }

    fn get_key(map: &HashMap<String, String>, primary: &str, secondary: &str) -> Option<String> {
        map.get(primary)
            .or_else(|| map.get(secondary))
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
    }

    CredentialBundle {
        openai_api_key: get_key(&env_map, "OPENAI_API_KEY", "OPENPLANTER_OPENAI_API_KEY"),
        anthropic_api_key: get_key(
            &env_map,
            "ANTHROPIC_API_KEY",
            "OPENPLANTER_ANTHROPIC_API_KEY",
        ),
        openrouter_api_key: get_key(
            &env_map,
            "OPENROUTER_API_KEY",
            "OPENPLANTER_OPENROUTER_API_KEY",
        ),
        cerebras_api_key: get_key(
            &env_map,
            "CEREBRAS_API_KEY",
            "OPENPLANTER_CEREBRAS_API_KEY",
        ),
        exa_api_key: get_key(&env_map, "EXA_API_KEY", "OPENPLANTER_EXA_API_KEY"),
        voyage_api_key: get_key(&env_map, "VOYAGE_API_KEY", "OPENPLANTER_VOYAGE_API_KEY"),
    }
}

/// Build credentials from process environment variables.
pub fn credentials_from_env() -> CredentialBundle {
    fn env_key(primary: &str, secondary: &str) -> Option<String> {
        env::var(primary)
            .ok()
            .or_else(|| env::var(secondary).ok())
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
    }

    CredentialBundle {
        openai_api_key: env_key("OPENPLANTER_OPENAI_API_KEY", "OPENAI_API_KEY"),
        anthropic_api_key: env_key("OPENPLANTER_ANTHROPIC_API_KEY", "ANTHROPIC_API_KEY"),
        openrouter_api_key: env_key("OPENPLANTER_OPENROUTER_API_KEY", "OPENROUTER_API_KEY"),
        cerebras_api_key: env_key("OPENPLANTER_CEREBRAS_API_KEY", "CEREBRAS_API_KEY"),
        exa_api_key: env_key("OPENPLANTER_EXA_API_KEY", "EXA_API_KEY"),
        voyage_api_key: env_key("OPENPLANTER_VOYAGE_API_KEY", "VOYAGE_API_KEY"),
    }
}

/// Discover `.env` file candidates by walking from workspace up to ancestors.
pub fn discover_env_candidates(workspace: &Path) -> Vec<PathBuf> {
    let ws = workspace
        .canonicalize()
        .unwrap_or_else(|_| workspace.to_path_buf());
    let mut candidates = Vec::new();
    let mut dir = Some(ws.as_path());
    while let Some(d) = dir {
        let env_path = d.join(".env");
        if env_path.exists() {
            candidates.push(env_path);
            break; // use the nearest .env
        }
        dir = d.parent();
    }
    candidates
}

/// Workspace-level credential store at `{workspace}/.openplanter/credentials.json`.
pub struct CredentialStore {
    pub credentials_path: PathBuf,
}

impl CredentialStore {
    pub fn new(workspace: &Path, session_root_dir: &str) -> Self {
        let ws = workspace
            .canonicalize()
            .unwrap_or_else(|_| workspace.to_path_buf());
        let root = ws.join(session_root_dir);
        let _ = fs::create_dir_all(&root);
        Self {
            credentials_path: root.join("credentials.json"),
        }
    }

    pub fn load(&self) -> CredentialBundle {
        let content = match fs::read_to_string(&self.credentials_path) {
            Ok(c) => c,
            Err(_) => return CredentialBundle::default(),
        };
        let payload: HashMap<String, serde_json::Value> = match serde_json::from_str(&content) {
            Ok(p) => p,
            Err(_) => return CredentialBundle::default(),
        };
        CredentialBundle::from_json(&payload)
    }

    pub fn save(&self, creds: &CredentialBundle) -> std::io::Result<()> {
        let payload = creds.to_json();
        if let Some(parent) = self.credentials_path.parent() {
            fs::create_dir_all(parent)?;
        }
        let json = serde_json::to_string_pretty(&payload)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
        fs::write(&self.credentials_path, json)?;
        // Set permissions to owner-only (0o600)
        #[cfg(unix)]
        {
            let perms = fs::Permissions::from_mode(0o600);
            let _ = fs::set_permissions(&self.credentials_path, perms);
        }
        Ok(())
    }
}

/// User-level credential store at `~/.openplanter/credentials.json`.
pub struct UserCredentialStore {
    pub credentials_path: PathBuf,
}

impl UserCredentialStore {
    pub fn new() -> Self {
        let home = env::var("HOME")
            .or_else(|_| env::var("USERPROFILE"))
            .unwrap_or_else(|_| ".".to_string());
        Self {
            credentials_path: PathBuf::from(home)
                .join(".openplanter")
                .join("credentials.json"),
        }
    }

    pub fn load(&self) -> CredentialBundle {
        let content = match fs::read_to_string(&self.credentials_path) {
            Ok(c) => c,
            Err(_) => return CredentialBundle::default(),
        };
        let payload: HashMap<String, serde_json::Value> = match serde_json::from_str(&content) {
            Ok(p) => p,
            Err(_) => return CredentialBundle::default(),
        };
        CredentialBundle::from_json(&payload)
    }

    pub fn save(&self, creds: &CredentialBundle) -> std::io::Result<()> {
        let payload = creds.to_json();
        if let Some(parent) = self.credentials_path.parent() {
            fs::create_dir_all(parent)?;
        }
        let json = serde_json::to_string_pretty(&payload)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
        fs::write(&self.credentials_path, json)?;
        #[cfg(unix)]
        {
            let perms = fs::Permissions::from_mode(0o600);
            let _ = fs::set_permissions(&self.credentials_path, perms);
        }
        Ok(())
    }
}

impl Default for UserCredentialStore {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_credential_bundle_default_has_none() {
        let bundle = CredentialBundle::default();
        assert!(!bundle.has_any());
    }

    #[test]
    fn test_credential_bundle_has_any() {
        let mut bundle = CredentialBundle::default();
        bundle.openai_api_key = Some("sk-test".into());
        assert!(bundle.has_any());
    }

    #[test]
    fn test_credential_bundle_merge_missing() {
        let mut a = CredentialBundle {
            openai_api_key: Some("existing".into()),
            ..Default::default()
        };
        let b = CredentialBundle {
            openai_api_key: Some("should-not-overwrite".into()),
            anthropic_api_key: Some("new-key".into()),
            ..Default::default()
        };
        a.merge_missing(&b);
        assert_eq!(a.openai_api_key, Some("existing".into()));
        assert_eq!(a.anthropic_api_key, Some("new-key".into()));
    }

    #[test]
    fn test_to_json_round_trip() {
        let bundle = CredentialBundle {
            openai_api_key: Some("sk-123".into()),
            anthropic_api_key: None,
            openrouter_api_key: Some("or-456".into()),
            ..Default::default()
        };
        let json = bundle.to_json();
        assert_eq!(json.get("openai_api_key").unwrap(), "sk-123");
        assert!(!json.contains_key("anthropic_api_key"));
        assert_eq!(json.get("openrouter_api_key").unwrap(), "or-456");
    }

    #[test]
    fn test_parse_env_file() {
        let dir = tempfile::tempdir().unwrap();
        let env_path = dir.path().join(".env");
        fs::write(
            &env_path,
            r#"
# Comment line
OPENAI_API_KEY=sk-from-env
export ANTHROPIC_API_KEY='ant-key'
EXA_API_KEY="exa-quoted"
UNRELATED_VAR=foo
"#,
        )
        .unwrap();

        let bundle = parse_env_file(&env_path);
        assert_eq!(bundle.openai_api_key, Some("sk-from-env".into()));
        assert_eq!(bundle.anthropic_api_key, Some("ant-key".into()));
        assert_eq!(bundle.exa_api_key, Some("exa-quoted".into()));
        assert!(bundle.cerebras_api_key.is_none());
    }

    #[test]
    fn test_credential_store_save_load() {
        let dir = tempfile::tempdir().unwrap();
        let store = CredentialStore::new(dir.path(), ".openplanter");
        let bundle = CredentialBundle {
            openai_api_key: Some("sk-test".into()),
            anthropic_api_key: Some("ant-test".into()),
            ..Default::default()
        };
        store.save(&bundle).unwrap();
        let loaded = store.load();
        assert_eq!(loaded.openai_api_key, Some("sk-test".into()));
        assert_eq!(loaded.anthropic_api_key, Some("ant-test".into()));
    }

    #[test]
    fn test_credential_store_load_missing() {
        let dir = tempfile::tempdir().unwrap();
        let store = CredentialStore::new(dir.path(), ".openplanter");
        let loaded = store.load();
        assert!(!loaded.has_any());
    }

    #[test]
    fn test_strip_quotes_fn() {
        assert_eq!(strip_quotes("'hello'"), "hello");
        assert_eq!(strip_quotes("\"world\""), "world");
        assert_eq!(strip_quotes("no-quotes"), "no-quotes");
        assert_eq!(strip_quotes("  'spaced'  "), "spaced");
    }
}
